### PR TITLE
Update cluster status in den If cluster.teardown() is called

### PR DESCRIPTION
**Pull request by BelSasha**

### CODE REVIEW

**Comment by BelSasha** [comment]
* **#837** <a href="https://app.graphite.dev/github/pr/run-house/runhouse/837?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @BelSasha and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: runhouse/resources/hardware/on_demand_cluster.py [hunk 1]
```diff
@@ -464,6 +467,38 @@ def teardown(self):
         Example:
             >>> rh.ondemand_cluster("rh-cpu").teardown()
         """
+        # TODO [SB]: remove the den_auth check once we will get status of clusters without den_auth as well.
+        if self.den_auth:
+            try:
+                # cluster_config = self.config()
+                cluster_status_data = self.status()
+                status_data = {
+                    "status": "terminated",
```
Comment (jlewitt1) [inline runhouse/resources/hardware/on_demand_cluster.py:476]: do we have an enum for this?

File: runhouse/resources/hardware/on_demand_cluster.py [hunk 2]
```diff
@@ -464,6 +467,38 @@ def teardown(self):
...
+                    "resource_type": cluster_status_data.get("cluster_config").get(
```
Comment (jlewitt1) [inline runhouse/resources/hardware/on_demand_cluster.py:477]: do we have access to the cluster config separately here? meaning it seems like its embedded inside the cluster status data, is it the exact same data?

File: runhouse/resources/hardware/on_demand_cluster.py [hunk 3]
```diff
@@ -464,6 +467,38 @@ def teardown(self):
...
+                )
+                post_status_data_resp = requests.post(
+                    f"{api_server_url}/resource/{cluster_uri}/cluster/status",
+                    data=json.dumps(status_data),
+                    headers=rns_client.request_headers(),
+                )
+                assert post_status_data_resp.status_code == 200
+            except Exception:
+                warning_msg = "Failed to update in Den that the cluster is terminated."
+                if post_status_data_resp:
```
Comment (jlewitt1) [inline runhouse/resources/hardware/on_demand_cluster.py:494]: for this try / except  - the handling of an exception in the logic here can be separated from  response with a non 200 status code, but right now they look bundled together

File: runhouse/resources/hardware/on_demand_cluster.py [hunk 4]
```diff
@@ -464,6 +467,38 @@ def teardown(self):
...
+                    "api_server_url", rns_client.api_server_url
+                )
+                post_status_data_resp = requests.post(
+                    f"{api_server_url}/resource/{cluster_uri}/cluster/status",
+                    data=json.dumps(status_data),
+                    headers=rns_client.request_headers(),
+                )
+                assert post_status_data_resp.status_code == 200
+            except Exception:
+                warning_msg = "Failed to update in Den that the cluster is terminated."
```
Comment (jlewitt1) [inline runhouse/resources/hardware/on_demand_cluster.py:493]: any specific motivation for storing the warning text as a separate variable?

File: tests/test_resources/test_clusters/test_on_demand_cluster.py [hunk 5]
```diff
@@ -59,3 +61,27 @@ def test_restart_does_not_change_config_yaml(self, cluster):
             assert config_yaml_res_after_restart[0][0] == 0
             config_yaml_content_after_restart = config_yaml_res[0][1]
             assert config_yaml_content_after_restart == config_yaml_content
+
+    @pytest.mark.level("minimal")
+    def test_set_status_after_teardown(self, ondemand_aws_https_cluster_with_auth):
```
Comment (jlewitt1) [inline tests/test_resources/test_clusters/test_on_demand_cluster.py:66]: how come we're using a hard coded cluster fixture here?

File: runhouse/resources/hardware/cluster.py [hunk 6]
```diff
@@ -717,7 +717,7 @@ def status(self, resource_address: str = None):
             return status
 
         except ValueError as e:
-            raise e
```
Comment (jlewitt1) [inline runhouse/resources/hardware/cluster.py:720]: don't we want to raise the original exception here?

File: runhouse/resources/hardware/on_demand_cluster.py (hunk 1 above)
Comment (BelSasha) [reply runhouse/resources/hardware/on_demand_cluster.py:476]: We have one on the den side, will add same in OSS as well.

File: runhouse/resources/hardware/on_demand_cluster.py (hunk 2 above)
Comment (BelSasha) [reply runhouse/resources/hardware/on_demand_cluster.py:477]: It mostly the same data. The cluster_config embedded inside the cluster status contains the following additional keys: domain and server_host. 

The only reason we need to access the cluster_config is to get the resource_type, but we can extract it using `self.__class__.__base__.__name__ `(for sub-resources of cluster) or` self.__class__.__name__` (for clusters)

File: runhouse/resources/hardware/on_demand_cluster.py (hunk 4 above)
Comment (BelSasha) [reply runhouse/resources/hardware/on_demand_cluster.py:493]: Just for the sake of code reusability: when handling an error where status_code != 200 or when catching an exception, the prefix of the warning message is the same, but the suffix is different.

Don't have strong opinion on this though, can remove this variable.

File: runhouse/resources/hardware/on_demand_cluster.py [hunk 7]
```diff
@@ -464,6 +467,36 @@ def teardown(self):
         Example:
             >>> rh.ondemand_cluster("rh-cpu").teardown()
         """
+        # TODO [SB]: remove the den_auth check once we will get status of clusters without den_auth as well.
+        warning_msg = "Failed to update in Den that the cluster is terminated."
```
Comment (jlewitt1) [inline runhouse/resources/hardware/on_demand_cluster.py:471]:
```suggestion
        warning_msg = "Failed to update Den with cluster terminated status."
```

File: runhouse/resources/hardware/on_demand_cluster.py [hunk 8]
```diff
@@ -464,6 +467,36 @@ def teardown(self):
...
+        if self.den_auth:
+            try:
+                cluster_status_data = self.status()
+                status_data = {
+                    "status": ResourceServerStatus.terminated,
+                    "resource_type": self.__class__.__base__.__name__.lower(),
```
Comment (jlewitt1) [inline runhouse/resources/hardware/on_demand_cluster.py:477]: i think we can do: `self.__class__.__name__.lower()`? Also i think we use this elsewhere to track the `resource_subtype`, might be worth creating a property for that

File: tests/test_resources/test_clusters/test_on_demand_cluster.py [hunk 9]
```diff
@@ -155,3 +157,28 @@ def test_autostop(self, cluster):
...
+            f"{api_server_url}/resource/{cluster_uri}/cluster/status",
+            headers=rns_client.request_headers(),
+        )
+
+        assert get_status_data_resp.status_code == 200
+        # For UI displaying purposes, the cluster/status endpoint returns cluster status history.
+        # The latest status info is the first element in the list returned by the endpoint.
+        get_status_data = get_status_data_resp.json()["data"][0]
+        assert get_status_data["resource_type"] == cluster_config.get("resource_type")
+        assert get_status_data["status"] == "terminated"
```
Comment (jlewitt1) [inline tests/test_resources/test_clusters/test_on_demand_cluster.py:184]: can use the enum here too

**Pull request merged**